docs/proposals/persistent-storage-provisioning.md

```diff
@@ -0,0 +1,58 @@
+# Persistent Storage Provisioning
+
+This document proposes a design taking advantage of cloud provider APIs to automate provisioning of persistent storage.
+
+
+## Background
+
+### Kubernetes Infrastructure (current and under development).
+
+Kubernetes currently has a mechanism for users to
+[request storage](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/persistent-storage.md#request-storage)
+for their pods, and a mechanism for admins to
+[describe the storage](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/persistent-storage.md#describe-available-storage)
+they have created and made available for use by pods. The various
+mechanisms for mounting the provided storage are provided by plugins
+such as the
+[GCE PD](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/volume/gce_pd/gce_pd.go),
+[NFS plugin](https://github.com/GoogleCloudPlatform/kubernetes/pull/4601),
+the [iSCSI plugin](https://github.com/GoogleCloudPlatform/kubernetes/pull/4612)
+
+For more detail please see the [Persistent Storage design document](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/persistent-storage.md)
+
+### Cloud Provider APIs
+
+Cloud providers usually provide an API for requesting storage dynamically.
+
+For example:
+
+GCE: https://cloud.google.com/compute/docs/reference/latest/disks/insert
+
+AWS: http://docs.aws.amazon.com/AmazonS3/latest/API/SOAPCreateBucket.html
+
+Rackspace: http://api.rackspace.com/api-ref-blockstorage.html#volumes
+
+
+## Proposal
+
+With all the building blocks mentioned above in place it is now possible to add functionality to use cloud APIs to
+automatically provide this storage as needed. The goal of this proposal is to reduce the work required of administrator to
+specifying a description of the persistent volumes needed. Kubernetes will then know which APIs to use to provision storage.
+
+To do this I propose adding the following components:
+
+**'PersistentVolumeProvider'**: A new interface which requires Create and Recycle methods.
+The actual implementation of using the APIs to request the storage form the cloud provider and announce its availability will be
+delegated to cloud specific plugins in a manner similar to how the persistent volume mounting is delegated to volume type
+specific implementations.
+
+**'PersistentVolumeController'**: A new API type analogous to a 'replicationControler'. A 'PersistentVolumeController' uses
+a the 'PersistentVolumeProvier' to ensure that a specified number of persistent volumes is created and ready to be matched to
+claims. As claims get matched the controller creates more volumes to ensure the requested level is maintained in the pool.
+The controller also watches for claims deleted by users -marked 'Recycled'- and takes care of recycling and returning those
+volumes to the pool.
```

Comment:
> I don't think it's the creator of the pod who would create pvControllers. This is still the domain of the admin provisioning the cluster. The end user knows nothing about the storage infrastructure underlying the cluster.

Hmm.. in that case -if there is no communication between the user and the admin of the cluster- we need to handle the situation where the admin creates one type/size of volume and claims are seeking a different one. Perhaps it is outside the scope of this proposal but I think there is merit there. Dynamic creation can be the next fall back if nothing suitable is found in the pools provided by the existing 'PersistentVolumeControllers'

> If we also support dynamic creation, then the provider has to know if it can create that resource on-the-fly. Some providers might not support this, but if it does, we need to match the claim to a provider that can create it.

Yeah, perhaps the implementation for a particular provider can also include a description of its capabilities. 

> there's a race condition now in that any other claim matching the newly created volume could be bound to it. A suitable volume would eventually be created and match the claim.

Yes this needs to be handled with care because we don't want a 5GB claim to wait unmatched while 100GB volume is being provisioned. We also don't want a 100GB persistent volume created to satisfy a 100GB claim be taken up by a newly arrived 5GB claim
Yeah we would need to handle that